### PR TITLE
[dynamo 3.11] refactor cpython function defs out of eval_frame.c

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99180
* #99934
* __->__ #99947

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @soumith @anijain2305 @Xia-Weiwen @desertfire

### CODE REVIEW

**Review by voznesenskym — APPROVED** [review]
Looks like all code motion, stamped. Please make sure CI passes.

File: torch/csrc/dynamo/cpython_defs.c [hunk 1]
```diff
@@ -0,0 +1,137 @@
+#include "cpython_defs.h"
```
Comment (voznesenskym) [inline torch/csrc/dynamo/cpython_defs.c:1]: Naming nit - `torch/csrc/dynamo/cpython_defs.c` is a weird file name, but what we are doing here is also weird.

File: torch/csrc/dynamo/cpython_defs.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+#if IS_PYTHON_3_11_PLUS
+
+#include <internal/pycore_frame.h>
+
+static int
+_PyFrame_OpAlreadyRan(_PyInterpreterFrame *frame, int opcode, int oparg);
+
+int
+THP_PyFrame_FastToLocalsWithError(_PyInterpreterFrame *frame);
```
Comment (voznesenskym) [inline torch/csrc/dynamo/cpython_defs.h:15]: Wouldn't hurt to link to copied-from source, ex: https://github.com/python/cpython/blob/main/Objects/frameobject.c#L1103

Comment (albanD) [reply torch/csrc/dynamo/cpython_defs.h:15]: There is a detailed comment and link in the .c file
Do you think the header is a better place to put it?

**Review by albanD — APPROVED** [review]
You're missing a definition in build_variables.bzl to appease compilation. You can grep for eval_frame.c in our build files and add this one next to it.

Sounds good once the build is passing

**Review by albanD — APPROVED** [review]
Looks good!

**Pull request closed without merging**